### PR TITLE
update the file cache within the write lock

### LINKED ISSUE #13391: Data loss on rename of a 49 GB folder
I accidentially renamed a folder on my production instance:

* the folder was filled up with nearly 20k files in a total size of 48,7 GB
* I renamed the folder in the web UI
* shortly (2-5 seconds) after I noticed this I shut down the client to avoid bigger trouble (as this folder was set up as a synced folder inside the client)
* the spinner spinned forever

Notes
* I have a database dump here
* I have all apache logs here
* I have the owncloud.log here
* I have a filesystem snapshot

If someone wants to help me with digging in the debris is welcome.

Access log:

The rename:
```
127.0.0.1 - - [15/Jan/2015:13:32:39 +0100] "GET /index.php/apps/files/ajax/rename.php?dir=%2F&newname=Bildersd&file=Bilder HTTP/1.1" 503 1216
```

The access log filtered for the folder Bilder:
```
...
127.0.0.1 - - [15/Jan/2015:13:24:36 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:25:04 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:25:34 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:26:06 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:26:39 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 16662
127.0.0.1 - - [15/Jan/2015:13:27:36 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:28:06 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:28:36 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:29:06 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:29:36 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:30:05 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:30:34 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:31:04 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:31:36 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 411
127.0.0.1 - - [15/Jan/2015:13:32:09 +0100] "PROPFIND /remote.php/webdav/Bilder HTTP/1.1" 207 16662
```

Nothing special in php-fpm.log or apache error log.

The folder is successfully renamed (in database and in the filesystem), but all database entries are gone (files are still there in the filesystem). Just a forced rescan with the occ command line tool was able to get them back into the database. Browsing the folder in the web UI didn't trigger the update of the file cache.

! For the user (without admin rights) it's not possible to get back the data from the server. It's simply not shown.

I will try to investigate further and try to reproduce. 

cc @karlitschek @DeepDiver1975 FYI could get a showstopper soon - I opened this ticket to document my process

### PR DESCRIPTION
**Pull request by icewind1991**
Prevents any other process from updating the cache while we're working on it.

Fixes #13391

With just this PR trying to list the folder being renamed will result in a locked exception being thrown and the sync client will try again later, together with #16963 you can still browser the folder being renamed (actually trying to download the file will result in a locked error) and the sync client wont detect any changes until the rename is done.

cc @PVince81 @DeepDiver1975

### CODE REVIEW

File: lib/private/files/cache/scanner.php [hunk 1]
```diff
@@ -124,13 +124,17 @@ public function getData($path) {
 	 * @param int $reuseExisting
 	 * @param int $parentId
 	 * @param array | null $cacheData existing data in the cache for the file to be scanned
+	 * @param bool $lock
```
Comment (PVince81) [inline lib/private/files/cache/scanner.php]: `false to disable cache locking` ?

File: lib/private/files/cache/scanner.php [hunk 2]
```diff
@@ -124,13 +124,17 @@ public function getData($path) {
...
 	 * @return array an array of metadata of the scanned file
+	 * @throws \OC\ServerNotAvailableException
```
Comment (PVince81) [inline lib/private/files/cache/scanner.php]: also throws LockedException ?

**Comment by PVince81** [comment]
This PR alone doesn't fully fix https://github.com/owncloud/core/issues/13391, but in my tests the folder contents was not completely removed. However, after a few tries a few files went missing.

Now testing together with https://github.com/owncloud/core/pull/16963 as advised in the original comment.

**Comment by PVince81** [comment]
Ran the rename test case 10 times with this PR merged together with https://github.com/owncloud/core/pull/16963 and the data loss seems to have disappeared!
I even had a case where I renamed twice before the sync client could catch up, still worked fine.

@icewind1991 you rock :smile: 

How about just merging into this PR here to make it easier for regression testing ?

I think we cannot merge this yet because there is still this regression https://github.com/owncloud/core/pull/16963#issuecomment-112795034 (but I'm ok if we do merge and fix it separately)

**Comment by PVince81** [comment]
:+1: (please fix the comments)

**Comment by icewind1991** [comment]
> (please fix the comments)

Already done :smile:

**Comment by icewind1991** [comment]
@DeepDiver1975 @MorrisJobke please review

**Comment by PVince81** [comment]
Tested both PRs merged, fixes https://github.com/owncloud/core/issues/11795 too

**Comment by PVince81** [comment]
Tested both PRs merged, fixes https://github.com/owncloud/core/issues/16584 too

**Comment by nickvergessen** [comment]
:+1: tested together with https://github.com/owncloud/core/pull/16963#issuecomment-113206603

**Pull request merged**